asf_welsh_energy_consultation/pipeline/process_data.py

```diff
@@ -384,3 +384,122 @@ def generate_age_data(wales_df):
     age_data["cumul_prop"] = age_data["percentage"].cumsum()
 
     return age_data
+
+
+def get_installations_per_year():
+    """
+    Get MCS installations per year for Wales.
+    Returns:
+        pandas.DataFrame of MCS installations per year in Wales.
+
+    """
+    mcs = get_enhanced_mcs()
+    mcs["n"] = 1
+    mcs["year"] = pd.to_datetime(mcs["commission_date"]).dt.year
+    installations_by_year = mcs.groupby("year")["n"].sum().reset_index()
+
+    # Sort by date ascending
+    installations_by_year = installations_by_year.sort_values("year")
+    installations_by_year = installations_by_year.rename(
+        columns={"commission_date": "date"}
+    )
+
+    return installations_by_year
+
+
+def mean_installations_per_year(min_year, max_year):
+    """
+    Get mean average MCS installations in Wales per year for given date range.
+    Args:
+        min_year: Minimum year (exclusive)
+        max_year: Maximum year (exclusive)
+
+    Returns:
+        int: Mean average MCS installations per year.
+    """
+    installations_by_year = get_installations_per_year()
+    subset = installations_by_year[
+        (installations_by_year["year"] > min_year)
+        & (installations_by_year["year"] < max_year)
+    ]
+
+    return subset["n"].mean()
+
+
+def median_installations_per_year(min_year, max_year):
+    """
+    Get median average MCS installations in Wales per year for given date range.
+    Args:
+        min_year: Minimum year (exclusive)
+        max_year: Maximum year (exclusive)
+
+    Returns:
+        int: Median average MCS installations per year.
+    """
+    installations_by_year = get_installations_per_year()
+    subset = installations_by_year[
+        (installations_by_year["year"] > min_year)
+        & (installations_by_year["year"] < max_year)
+    ]
+
+    return subset["n"].median()
+
+
+def get_total_rural_and_urban_properties():
+    """
+    Get total count of properties in Wales in urban vs rural locations.
+
+    Returns:
+        dict: Percent of rural and urban properties in Wales.
+    """
+    rural = get_data.get_rurality()
+    dwellings = get_data.get_dwelling_data()
+
+    df = dwellings.merge(rural, how="left", on="lsoa_code")
+    df = df[df["country"] == "W"]
+    rurality = df.groupby("rural_2")["total_dwellings"].sum().to_dict()
+    rurality_pct_dict = {
+        "Rural": (rurality["Rural"] / (rurality["Rural"] + rurality["Urban"])) * 100,
+        "Urban": (rurality["Urban"] / (rurality["Rural"] + rurality["Urban"])) * 100,
+    }
+
+    return rurality_pct_dict
+
+
+def get_total_on_off_gas_postcodes():
+    """
+    Get total count of postcodes in Wales which are on- vs off-gas.
+
+    Returns:
+        dict: Percent of on- and off-gas postcodes in Wales.
+    """
+
+    og = get_data.get_offgas()
```

Comment:
This data is for all the countries, not just wales right?

Reply:
Yes. From Xoserve website: 'Off-gas refers to postcode areas in Great Britain where we don't have any gas meter points on the UK Link System.'